### DISCUSSION TITLE
Dynamic scopes and filters

### DISCUSSION THREAD

**Discussion by seanlinsley** [activeadmin/activeadmin#7862] [Ideas]
It's a common misconception (#2340, #2781) that scopes will behave dynamically, but in reality they're loaded once, when the server boots.

I'm open to DSL suggestions.

**Comment by rraub** [comment]
If the common misconception is that they are dynamic, then why not change default to dynamic and make the static scopes declared explicitly?

```
scope :something, static: true
```

or 

```
static_scope :something
```

**Comment by seanlinsley** [comment]
It's not that simple because people expect this to be dynamic:

``` ruby
ActiveAdmin.register User do

  User.pluck(:roles).each do |role|
    scope(role) { |scope| scope.where role: role }
  end
```

When it can't possibly be since the `register` block is only run once.

**Comment by Fivell** [comment]
maybe  it would be better to create new method for dynamic scopes

``` ruby
ActiveAdmin.register User do

  scope :all
  scope :active
  scope :disactive
 # this block runs every request
  within_scopes_navigation do 
    User.pluck(:roles).each do |role|
       scope(role) { |scope| scope.where role: role }
    end
  end
end
```

**Comment by shekibobo** [comment]
``` ruby
generate_scopes do
  User.pluck(:roles).each do |role|
    scope(role) { |scope| scope.where(role: role) }
  end
end
```

Seems like putting it in a block to dynamically generate them would be the best DSL, whatever the method is named. It keeps it flexible enough to let the user make them however they need to.

`generate_scopes`, `dynamic_scopes`, `render_scopes`, wherein calling `scope` will add another scope to the list.

I don't know if it would make sense or impact performance too much, but what if we did:

``` ruby
scope :all # deprecated

# Everything goes inside the +scopes+ block
scopes do
  scope :all
  scope :active
  scope :inactive
  User.pluck(:roles).each do |role|
    scope(role) { |users| user.where(role: role) }
  end
end

# Accepts a list of defined scopes on the resource as well as a block for custom scopes
scopes :all, :active, :inactive do
  scope(:custom_scope) { |users| users.where(something: true) }

  User.pluck(:roles).each do |role|
    scope(role) { |users| user.where(role: role) }
  end

  # why not be in the view context
  controller.parent.things.each do |thing|
    scope(thing) { |users| users.includes(:parent).where(parent: { thing: thing }) }
  end
end
```

In this case, we enforce keeping all the scopes in one place. If we enforce putting the all `scope` calls to be inside the `scopes` block, couldn't the `scopes` method just be an arbre context that opens up inside the index pages? That'll give greater access to things in the current session (like the parent of a nested resource), and we don't have to reopen the page to try to figure out where to rerender new dynamic scopes.

**Comment by rraub** [comment]
I like @shekibobo 's scopes suggestion. 

What about filters? I could see a desire for a similar behavior with them. 

```
filters do 
  filter :handy_filter if user.is_able? :see_handy_filter
end
```

you could even change the filters based on the current scope..

**Comment by shekibobo** [comment]
@rraub I was thinking that about half way through writing it.

**Comment by timoschilling** [comment]
I prefer @shekibobo's

``` ruby
scope :all # deprecated

# Everything goes inside the +scopes+ block
scopes do
  scope :all
  scope :active
  scope :inactive
  User.pluck(:roles).each do |role|
    scope(role) { |users| user.where(role: role) }
  end
end
```

and we should build @rraub's suggestion to:

``` ruby
filters do 
  filter :handy_filter if user.is_able? :see_handy_filter
end
```

**Comment by itkin** [comment]
+1

**Comment by heaven** [comment]
Any luck with this?

**Comment by gingerlime** [comment]
also wondering about this... in our particular case, we have a "heavy" scope, which I only want to run when it's explicitly needed. Otherwise, the index page would run a `count` on the scope to show how many entries there are, but I'd rather avoid it and only apply it with an action item or something... Would be great to be able to apply the scope only with the scope param, but this doesn't seem to work if the scope isn't specified.
